<?xml version="1.0"?>
<launch>
  <!-- Arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="open_rviz" default="true"/>
  <arg name="map_size" default="384"/> <!-- Fixed map size in pixels -->
  <arg name="map_resolution" default="0.05"/> <!-- Map resolution in meters/pixel -->
  
  <!-- Map size calculation -->
  <arg name="map_width_meters" value="$(eval map_size * map_resolution)"/>
  <arg name="map_height_meters" value="$(eval map_size * map_resolution)"/>
  
  <!-- Turtlebot3 -->
  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)" />
  </include>

  <!-- SLAM: Hector -->
  <node pkg="hector_mapping" type="hector_mapping" name="hector_mapping" output="screen">
    <!-- Frame names -->
    <param name="map_frame" value="map" />
    <param name="base_frame" value="base_footprint" />
    <param name="odom_frame" value="odom" />
    
    <!-- Map size / start point -->
    <param name="map_resolution" value="$(arg map_resolution)"/>
    <param name="map_size" value="$(arg map_size)"/>
    <param name="map_start_x" value="0.5"/>
    <param name="map_start_y" value="0.5"/>
    <param name="map_multi_res_levels" value="2" />
    
    <!-- Map update parameters -->
    <param name="update_factor_free" value="0.4"/>
    <param name="update_factor_occupied" value="0.9" />    
    <param name="map_update_distance_thresh" value="0.4"/>
    <param name="map_update_angle_thresh" value="0.06" />
    <param name="pub_map_odom_transform" value="true"/>
    
    <!-- Laser scan parameters -->
    <param name="scan_subscriber_queue_size" value="5"/>
    <param name="scan_topic" value="scan"/>
    <param name="use_tf_scan_transformation" value="true"/>
    <param name="use_tf_pose_start_estimate" value="false"/>
    
    <!-- Debug parameters -->
    <param name="pub_drawings" value="true"/>
    <param name="pub_debug_output" value="true"/>
  </node>
  
  <!-- RVIZ -->
  <group if="$(arg open_rviz)"> 
    <node pkg="rviz" type="rviz" name="rviz" required="true"
          args="-d $(find turtlebot3_slam)/rviz/turtlebot3_hector.rviz"/>
  </group>
</launch>